entc/gen/globalid.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2019-present Facebook Inc. All rights reserved.
+// This source code is licensed under the Apache 2.0 license found
+// in the LICENSE file in the root directory of this source tree.
+
+package gen
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"entgo.io/ent/dialect/entsql"
+)
+
+const (
+	incrementIdent = "const IncrementStarts"
+	rangeSize      = 1 << 32
+)
+
+// IncrementStarts holds the autoincrement start value for each type.
+type IncrementStarts map[string]int64
+
+// IncrementStartAnnotation assigns a unique range to each node in the graph.
+func IncrementStartAnnotation(g *Graph) error {
+	// To ensure we keep the existing type ranges, load the current global id sequence, if there is any.
+	var (
+		r        = make(IncrementStarts)
+		path     = rangesFilePath(g.Target)
+		buf, err = os.ReadFile(path)
+	)
+	switch {
+	case os.IsNotExist(err): // first time generation
+	case err != nil:
+		return err
+	default:
+		var (
+			matches = make([][]byte, 0, 2)
+			lines   = bytes.Split(buf, []byte("\n"))
+		)
+		for i := 0; i < len(lines); i++ {
+			if l := lines[i]; bytes.HasPrefix(l, []byte(incrementIdent)) {
+				matches = append(matches, l)
+			}
+		}
+		if len(matches) != 1 {
+			return fmt.Errorf("expect to have exactly 1 ranges in %s", path)
+		}
+		var (
+			line  = matches[0]
+			start = bytes.IndexByte(line, '"')
+			end   = bytes.LastIndexByte(line, '"')
+		)
+		if start == -1 || start >= end {
+			return fmt.Errorf("unexpected line %s", line)
+		}
+		l, err := strconv.Unquote(string(line[start : end+1]))
+		if err != nil {
+			return err
+		}
+		if err := json.Unmarshal([]byte(l), &r); err != nil {
+			return fmt.Errorf("unmarshal ranges: %w", err)
+		}
+	}
+	// Range over all nodes and assign the increment starting value.
+	var (
+		need []*Type
+		high int64
+	)
+	for _, n := range g.Nodes {
+		if n.Annotations == nil {
+			n.Annotations = make(Annotations)
+		}
+		a := n.EntSQL()
+		if a == nil {
+			a = &entsql.Annotation{}
+			n.Annotations[a.Name()] = a
+		}
+		switch v, ok := r[n.Table()]; {
+		case a.IncrementStart != nil:
+			// In case the start value is defined by an annotation already, it has precedence.
+			r[n.Table()] = *a.IncrementStart
```

Comment:
I believe we should validate the ranges. Protect the user from doing mistakes here. If global ID is enabled, two node types cannot have overlapped ranges:
<img width="1008" alt="image" src="https://github.com/user-attachments/assets/dbb1b074-3ce5-4454-acca-8da9a81e69f6" />

Reply:
WDYT, @giautm and @masseelch?

Reply:
How do you want to ensure there is no overlap? The only thing making slightly sense is to check two nodes have not the same increment start counter. But users  might decide a specific table only requires a range of 10 entries. 

Do you want to enforce that there is at least a gap of 1<<32 between nodes?

Reply:
> Do you want to enforce that there is at least a gap of 1<<32 between nodes?

I think so. This is how it works now.